crates/ruff_db/src/diagnostic/render.rs

```diff
@@ -820,6 +828,248 @@ fn relativize_path<'p>(cwd: &SystemPath, path: &'p str) -> &'p str {
     path
 }
 
+/// A measure of the width of a line of text.
+#[derive(Clone, Copy, Default)]
+struct LineWidthBuilder {
+    /// The width of the line.
+    width: usize,
+    /// The column of the line.
+    /// This is used to calculate the width of tabs.
+    column: usize,
+}
+
+impl LineWidthBuilder {
+    fn get(&self) -> usize {
+        self.width
+    }
+
+    /// Adds the given character to the line width.
+    #[must_use]
+    fn add_char(mut self, c: char) -> Self {
+        const TAB_SIZE: usize = 4;
+        match c {
+            '\t' => {
+                let tab_offset = TAB_SIZE - (self.column % TAB_SIZE);
+                self.width += tab_offset;
+                self.column += tab_offset;
+            }
+            '\n' | '\r' => {
+                self.width = 0;
+                self.column = 0;
+            }
+            _ => {
+                self.width += unicode_width::UnicodeWidthChar::width(c).unwrap_or(0);
+                self.column += 1;
+            }
+        }
+        self
+    }
+}
+
+/// Given some source code and annotation ranges, this routine replaces tabs
+/// with ASCII whitespace, and unprintable characters with printable
+/// representations of them.
+///
+/// The source code and annotations returned are updated to reflect changes made
+/// to the source code (if any).
+fn replace_whitespace_and_unprintable<'r>(
+    source: &'r str,
+    mut annotations: Vec<RenderableAnnotation<'r>>,
+) -> EscapedSourceCode<'r> {
+    let mut result = String::new();
+    let mut last_end = 0;
+    let original_ranges: Vec<TextRange> = annotations.iter().map(|ann| ann.range).collect();
+    let mut line_width = LineWidthBuilder::default();
```

Comment:
Nit: Move the declaration of this variable closer to where it's used. 

Same for `last_end`. It helps readability because, as a reader, I have to page less variables into my memory :)